### PR TITLE
memoize user's credentials and refactoring

### PR DESCRIPTION
**Pull request by dodikk**
* Added credentials memoization to the demo code. Addressing https://github.com/jivesoftware/JiveAuthenticatingHTTPProtocol/issues/6
* Some refactoring of the library code. Addressing https://github.com/jivesoftware/JiveAuthenticatingHTTPProtocol/issues/3
* Some readme additions based on https://github.com/jivesoftware/JiveAuthenticatingHTTPProtocol/issues/4 and https://github.com/jivesoftware/JiveAuthenticatingHTTPProtocol/issues/5 . Hopefully, it has become a bit prettier and easier to read.
* Lots of changes made by `pod install` command. Not changing the `.gitignore` policy, so I've committed them too.

Basically, the reviewer may ignore the contents of the `Pods` directory.

### CODE REVIEW

File: JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m [hunk 1]
```diff
@@ -9,142 +9,269 @@
...
+    [[UIAlertView alloc] initWithTitle: @"JAHPDemo"
+                               message: @"Enter 'foo' for the username and 'bar' for the password"
+                              delegate: self
+                     cancelButtonTitle: @"Cancel"
+                     otherButtonTitles: @"OK", nil];
+    
     self.authAlertView.alertViewStyle = UIAlertViewStyleLoginAndPasswordInput;
     [self.authAlertView show];
+    
+//    JAHPDidCancelAuthenticationChallengeHandler logCancelEvent =
```
Comment (hborders) [inline JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m]: Why did you comment this out?

File: JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m [hunk 2]
```diff
@@ -9,142 +9,269 @@
...
-                          otherButtonTitles:nil] show];
-    };
+    return result;
+}
+
+
+-(void)handleAuthChallenge:(NSURLAuthenticationChallenge*)challenge
+               forProtocol:(JAHPAuthenticatingHTTPProtocol*)authenticatingHTTPProtocol
+{
+    // TODO: maybe fix leaking `self`
```
Comment (hborders) [inline JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m]: Where is the leaking `self`?

**Review by hborders — CHANGES_REQUESTED** [review]
A lot of the PR refactors the example code, which I explicitly don't want, as I explained in #3.
Also, a lot of the PR changes existing formatting strategies. Please keep the original formatting.

File: JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m (hunk 1 above)
Comment (dodikk) [reply JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m]: That's what I've introduced before noticing the ```#define USE_PROTOCOL_FOR_CANCELLATION 0``` . Forgot to delete. Will fix that.

File: JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m (hunk 2 above)
Comment (dodikk) [reply JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m]: Same thing. Old irrelevant comment I've written before the refactoring happened. Will delete that.

File: JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m [hunk 3]
```diff
@@ -9,142 +9,261 @@
...
+    [[UIAlertView alloc] initWithTitle: @"JAHPDemo"
+                               message: @"Enter 'foo' for the username and 'bar' for the password"
+                              delegate: self
+                     cancelButtonTitle: @"Cancel"
+                     otherButtonTitles: @"OK", nil];
+    
     self.authAlertView.alertViewStyle = UIAlertViewStyleLoginAndPasswordInput;
     [self.authAlertView show];
+    
     return nil;
```
Comment (hborders) [inline JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m:92]: I'm fine with adding a logging challenge handler. I think that would make exploration a little simpler.

**Comment by hborders** [comment]
I'll have to do a more in-depth review later tonight.

File: JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m (hunk 3 above)
Comment (dodikk) [reply JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m:92]: I believe, if this function returns non-nil, the `authenticatingHTTPProtocol:didCancelAuthenticationChallenge:` method will end up not being called. Which is unwanted for the `#if USE_PROTOCOL_FOR_CANCELLATION` demo branch.

**Comment by hborders** [comment]
There's still a ton of noise in this PR from reformatting that isn't strictly necessary. If you want different formatting for this project, I'm fine with it, but let's make that a separate PR.

**Comment by dodikk** [comment]
> There's still a ton of noise in this PR from reformatting that isn't strictly necessary. If you want different formatting for this project, I'm fine with it, but let's make that a separate PR.

Ok. I understand. I'll try to split the changes later.

**Comment by dodikk** [comment]
Reverted original formatting in https://github.com/jivesoftware/JiveAuthenticatingHTTPProtocol/pull/9

Tested with 
* original example
* NTLM and "needs authorization video"

**Comment by dodikk** [comment]
Closing this since a more recent PR is available https://github.com/jivesoftware/JiveAuthenticatingHTTPProtocol/pull/9
